text/0000-grid-primitives.md

```diff
@@ -0,0 +1,618 @@
+    // Schema.
+    string name = 1;
+    // The data type of the property.  Indicates which value field the actual
+    // value may be found.  Must not be set to `UNSET_DATA_TYPE`.
+    PropertyDefinition.DataType data_type = 2;
+
+    // The value fields for the possible data types.  Only one of these will
+    // contain a value, determined by the value of `data_type`
+    bytes bytes_value = 10;
+    bool boolean_value = 11;
+    sint64 number_value = 12;
+    string string_value = 13;
+    uint32 enum_value = 14;
+    repeated PropertyValue struct_values = 15;
+}
+```
+
+## Property Examples
+
+The following examples show how these property definitions and instance values
+would be constructed using the above protobuf messages.  The examples use a
+Python-like syntax.
+
+### Bytes
+
+A bytes value would be represented as follows:
+
+```
+PropertyDefinition(
+    name="user_data",
+    data_type=PropertyDefinition.DataType.Bytes,
+    description="Arbitrary serialized user data."
+)
+```
+
+Because this is a protobuf message, the default value for this field is an empty
+byte array.
+
+### Boolean
+
+A boolean value would be represented as follows:
+
+```
+PropertyDefinition(
+    name="is_enabled",
+    data_type=PropertyDefinition.DataType.BOOLEAN,
+    required=True,
+    description="Indicates that the containing struct is enabled."
+)
+```
+
+The value would be represented as:
+
+```
+PropertyValue(
+    name="is_enabled",
+    data_type=PropertyDefinition.DataType.BOOLEAN,
+    boolean_value=True
+)
+```
+
+Because this is a protobuf message, the default value for this field is `False`.
+
+### String
+
+A UTF-8 string value would be represented as follows:
+
+```
+PropertyDefinition(
+    name="title",
+    data_type=PropertyDefinition.DataType.STRING,
+    required=True,
+    description="A blog post title."
+)
+```
+
+The value would be represented as:
+
+```
+PropertyValue(
+    name="title"
+    data_type=PropertyDefinition.DataType.STRING,
+    string_value="My Very Nice Blog Example"
+)
+```
+
+Because this is a protobuf message, the default value for this field is the
+empty string.
+
+### Number
+
+An integer value would be represented as the following type:
+
+```
+PropertyDefinition(
+    name="quantity",
+    data_type=PropertyDefinition.DataType.NUMBER,
+    number_exponent=0,
+    required=True,
+    description="The count of values in this container"
+)
+```
+
+This example shows an instance of a quantity of 23:
+
+```
+PropertyValue(
+    name="quantity",
+    data_type=PropertyDefinition.DataType.NUMBER,
+    number_value=23,
+)
+```
+
+A fractional value would be represented as the following type:
+
+```
+PropertyDefinition(
+    name="price",
+    data_type=PropertyDefinition.DataType.NUMBER,
+    number_exponent=-2,
+    required=True,
+    description="The the price this object"
+)
+
+```
+
+This example shows an instance of a price with the value $0.23:
```

Comment:
I think we need to account for a 'measure type' either as a struct example or as a declarative primitive / framework struct or something.  

A measure is a 'number' and a 'unit'.  This example is "price" which implies a unit (of a currency).  That is not a number, it's a measure of currency.

Reply:
The NUMBER is the fundamental data representation, not necessarily a unit. Measure type might make sense in NUMBER properties, but ultimately, it's just a string, since we probably shouldn't try to enumerate all possible measures. 

In the above, it could be "usd_cents", versus "euro_cents", etc

Reply:
I don't think you would really want to do "usd_cents" or "euro_cents". Using what is currently in this RFC, I think you would want to define a struct with two fields, one for denomination and the other for the number. Which at best makes this a poor example to have in the RFC.

I think it would be reasonable to add a CURRENCY DataType for the same reason we have LOCATION - because it's common and we want it standardized across grid. We could add it to this RFC, but if we can't agree how we should implement it (this is not a simple topic really), then we could mention it in the unresolved section of the document and add it with a different RFC following from further discussion.

Reply:
Added a paragraph to "Unresolved Questions" about promoting structs to first-class data types.

Reply:
My bias is to remove `LOCATION` under this same logic and also because there's more to location than just LAT/LONG.  If we rename to LAT_LONG or something then I would cede.

Reply:
I think the LAT_LONG naming would be good.  In other conversations, what constitutes a "location" seems to vary, in terms of data beyond lat/long.  With a lat/long value, any of those specific location types can be created as custom structs.

Reply:
Right on.

Reply:
Updated to be named LAT_LONG